features/activity/cancel_try_cancel/feature.java

```diff
@@ -0,0 +1,116 @@
+
+import io.temporal.activity.Activity;
+import io.temporal.activity.ActivityCancellationType;
+import io.temporal.activity.ActivityInterface;
+import io.temporal.activity.ActivityMethod;
+import io.temporal.client.ActivityCanceledException;
+import io.temporal.client.WorkflowClient;
+import io.temporal.common.RetryOptions;
+import io.temporal.failure.ActivityFailure;
+import io.temporal.failure.ApplicationFailure;
+import io.temporal.failure.CanceledFailure;
+import io.temporal.sdkfeatures.Feature;
+import io.temporal.sdkfeatures.Run;
+import io.temporal.sdkfeatures.Runner;
+import io.temporal.sdkfeatures.SimpleWorkflow;
+import io.temporal.workflow.*;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+@ActivityInterface
+public interface feature extends Feature, SimpleWorkflow {
+  @SignalMethod
+  void activityResult(String result);
+
+  @ActivityMethod
+  void cancellableActivity();
+
+  class Impl implements feature {
+    private CompletablePromise<String> activityResult;
+
+    @Override
+    public void workflow() {
+      activityResult = Workflow.newPromise();
+      var activities = activities(feature.class, builder -> builder
+              .setScheduleToCloseTimeout(Duration.ofMinutes(1))
+              .setHeartbeatTimeout(Duration.ofSeconds(5))
+              // Disable retry
+              .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(1).build())
+              .setCancellationType(ActivityCancellationType.TRY_CANCEL));
+
+      // Start cancellable activity
+      var activityPromise = new AtomicReference<Promise<Void>>();
+      var scope = Workflow.newCancellationScope(() ->
+              activityPromise.set(Async.procedure(activities::cancellableActivity))
+      );
+      scope.run();
+
+      // Sleep for short time (force task turnover)
+      Workflow.sleep(1);
+
+      // Cancel the activity and confirm it gets cancelled
+      scope.cancel();
+      try {
+        activityPromise.get().get();
+        throw ApplicationFailure.newFailure("No error", "NoError");
+      } catch (ActivityFailure e) {
+        if (!(e.getCause() instanceof CanceledFailure)) {
+          throw e;
+        }
+      }
+
+      // Confirm activity was cancelled
+      try {
+        var result = activityResult.get(10, TimeUnit.SECONDS);
+        if (!"cancelled".equals(result)) {
+          throw ApplicationFailure.newFailure("Expected cancelled, got: " + result, "BadResult");
+        }
+      } catch (TimeoutException e) {
+        throw ApplicationFailure.newFailureWithCause("Timeout", "Timeout", e);
+      }
+    }
+
+    @Override
+    public void activityResult(String result) {
+      activityResult.complete(result);
+    }
+
+    private WorkflowClient client;
+
+    @Override
+    public Run execute(Runner runner) throws Exception {
+      client = runner.client;
```

Comment:
This happened in Go and Java, might make sense to make it a thing available by default somehow?

Reply:
Yeah, perhaps, add a default activity interceptor in all langs to inject this into the activity context.

Reply:
I think we should avoid including any non-activity package things by default to a user (and even if we did, a default interceptor means we might as well hardcode it, an interceptor is an unnecessary abstraction). But for the harnesses here as a helper, sure.

IMO, a Temporal client is no more important than a DB client and the pattern to include X client should be the same for all values of X IMO.

Reply:
I disagree.
I actually think it's super useful in an activity for things like:
- query / signal my own workflow
- perform a checkpoint and wait for it to be committed (bypassing the heartbeat throttler)

Reply:
Of course it's useful. So are many other clients. I don't think it's any more useful than other clients though. If it's too difficult for them to get an external thing, be it our client or another client, we need to fix that. Not assume that the Temporal client is special. They have it at worker creation and activity registration time, there's no reason they can't make it available to the activity.